File: src/main/java/com/ampro/weebot/commands/games/Game.java
/**
 *
 */

package com.ampro.weebot.commands.games;

import com.ampro.weebot.bot.Weebot;
import net.dv8tion.jda.core.entities.User;

import java.util.ArrayList;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Basis of a Weebot Game.
 * Must be connected to a single type of Player.
 *
 * @param <P> A class that extends {@code Player}
 */
public abstract class Game<P extends Player> {

    /** Don't let some things be modified while the game is running */
    public static class ModificationWhileRunningException extends Exception {
        private static final long serialVersionUID = 1549072265432776147L;
        /** Parameterless constructor */
        public ModificationWhileRunningException() {}
        /** Constructor with message */
        public ModificationWhileRunningException(String err) { super(err); }

    }

    /** The ID of the hosting bot */
    protected final String HOST_ID;
    /** User ID of the User who started the game.*/
    protected final long AUTHOR_ID;
    //Keep a list of all the Players
    protected final ConcurrentHashMap<Long, P> PLAYERS;
    /** Is the game currently running? */
    protected boolean RUNNING;

    /** Create a game.
     *  Has empty Players list.
     * Starts not Running.
     * @param bot Weebot hosting the game
     */
    protected Game(Weebot bot, User author) {
        this.HOST_ID = bot.getBotId();
        this.RUNNING = false;
        this.PLAYERS = new ConcurrentHashMap<>();
        this.AUTHOR_ID = author.getIdLong();
        this.addUser(author);
    }

    /**
     * Create a game with an initial set of {@code Player}s
     * @param bot Weebot hosting game
     * @param players {@code Players} to add to the game
     */
    protected Game(Weebot bot, User author, P... players) {
        this.HOST_ID = bot.getBotId();
        this.RUNNING = false;
        this.PLAYERS = new ConcurrentHashMap<>();
        for (P p : players) {
            this.PLAYERS.putIfAbsent(p.getUser().getIdLong(), p);
        }
        this.AUTHOR_ID = author.getIdLong();
        this.addUser(author);
    }

    //Some Very important but vague methods to implement in child.
    protected abstract boolean startGame();
    protected abstract boolean endGame();

    /**
     * Add a user to the game, wrapping the {@link User} in a new {@link Player}
     * implementation.
     * @param user The user to add.
     * @return false if the user could not be added.
     */
    protected abstract boolean addUser(User user);

    /**
     * Add {@code Player} to the {@code Game}.
     * This method only attempts to add the player, any more action
     * should be implemented in a class-specific implementation.
     * @param player The player to add
     * @return  false if player is already in the Game. <br>
     *          true if player was added to the Game.
     */
    protected boolean joinGame(P player) {
        return this.PLAYERS.putIfAbsent(player.getUser().getIdLong(), player) == null
                   ? true : false;
    }

    /** Is the game ongoing? */
    public boolean isRunning() {
        return this.RUNNING;
    }

    /** Get an Iterable Arraylist of the players */
    public ArrayList<P> playerIterable() {
        return new ArrayList<>(this.PLAYERS.values());
    }

    public String getHOST_ID() {
        return HOST_ID;
    }

    public long getAUTHOR_ID() {
        return AUTHOR_ID;
    }

    public ConcurrentHashMap<Long, P> getPlayers() { return this.PLAYERS; }

    /**
     * Get a player.
     * @param user The user who's player to get.
     * @return The player or null if no player is found.
     */
    public P getPlayer(User user) {
        return this.PLAYERS.get(user.getIdLong());
    }

}

File: src/main/kotlin/com/ampro/weebot/commands/fun/games/Game.kt
/*
 * Copyright Aquatic Mastery Productions (c) 2018.
 */

package com.ampro.weebot.commands.`fun`.games

import com.ampro.weebot.database.constants.DEV_IDS
import net.dv8tion.jda.core.entities.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer

/**
 * File containing abstract classes and global variables used for Weebot Games.
 */

/** Don't let some things be modified while the game is running  */
class ModificationWhileRunningException : Exception {
    /** Constructor with message  */
    constructor(err: String) : super(err)

    companion object { private val serialVersionUID = 1549072265432776147L }
}

/**
 * A class representing a leaderboard
 *
 * @param scoredUsers A map of user IDs -> scores
 *
 * @author Jonathan Augustine
 * @since 2.0
 */
open class LeaderBoard(protected val scoredUsers: MutableList<Pair<Long, Int>>
                       = mutableListOf()) {
    fun addUser(user: User, score: Int = 0) {
        scoredUsers.add(Pair(user.idLong, score))
    }
    fun get() = scoredUsers.sortedBy { it.second }
    operator fun get(int: Int) = scoredUsers[int]
    operator fun get(user: User) = scoredUsers.firstOrNull { user.idLong == it.first }
}

/**
 * Base Wrapper Class for Members currently involved in a Webot `Game`
 *
 * @param user The [User] wrapped by this player object
 *
 * @author Jonathan Augustine
 * @since 1.0
 */
abstract class Player(/** User this Player is wrapped around*/ val user: User)
    : Comparable<Player> {

    /**
     * Send a private message to the `Player`.
     * @param message String to send
     */
    private fun privateMessage(message: String) {
        this.user.openPrivateChannel().queue { it.sendMessage(message).queue() }
    }

    /**
     * Send a private message to the [Player].
     *
     * @param message The message.
     * @param consumer Lambda
     */
    fun privateMessage(message: String, consumer: Consumer<Message>) {
        this.user.openPrivateChannel().queue {
            it.sendMessage(message).queue { m -> consumer.accept(m) }
        }
    }

    /**
     * Send a private embed message to the player.
     * @param embed The MessageEmbed to send.
     * @param consumer Message Consumer
     */
    fun privateMessage(embed: MessageEmbed, consumer: Consumer<Message>) {
        this.user.openPrivateChannel().queue {
            it.sendMessage(embed).queue { m -> consumer.accept(m) }
        }
    }

    override fun compareTo(other: Player): Int {
        return (this.user.idLong - other.user.idLong).toInt()
    }

}

/**
 * Under what conditions should the game end?
 *
 * @author Jonathan Augustine
 * @since 2.1
 */
enum class WinCondition(var score: Int) {
    /** End Game by Max Rounds */ ROUNDS(-1),
    /** End Game by Max Wins */ WINS(-1);
    operator fun invoke(score: Int) : WinCondition {
        this.score = score
        return this
    }
}

/**
 * Basis of a Weebot Game.
 * Must be connected to a single type of Player.
 *
 * @param P A class that extends [Player]
 */
abstract class Game<P : Player> (guild: Guild, author: User) {
    /** The ID of the hosting bot  */
    val guildID: Long = guild.idLong
    /** User ID of the User who started the game. */
    val authorID: Long = author.idLong
    /** List of all the Players */
    val players: ConcurrentHashMap<Long, P> = ConcurrentHashMap()
    /** Is the game currently running? Starts false */
    var isRunning: Boolean = false

    protected val playerList = mutableListOf<P>()

    init {
        this.addUser(author)
    }


    /**
     * Create a game with an initial set of `Player`s
     *
     * @param bot Weebot hosting game
     * @param players `Players` to add to the game
     */
    protected constructor(guild: Guild, author: User, vararg players: P)
            : this (guild, author) {
        players.forEach { addPlayer(it) }
        this.addUser(author)
    }

    //Some Very important but vague methods to implement in child.
    abstract fun startGame(): Boolean

    abstract fun endGame(): Boolean

    /**
     * Add a user to the game, wrapping the [User] in a new [Player] implementation.
     *
     * @param user The user to add.
     * @return false if the user could not be added.
     */
    abstract fun addUser(user: User) : Boolean

    /**
     * Adds a player to the [players] map and the [playerList]
     *
     * @return false if the player is already in the game
     */
    protected open fun addPlayer(player: P) = if (this.players[player.user.idLong] == null) {
        this.players[player.user.idLong] = player
        this.playerList.add(player)
        true
    } else false

    /**
     * Get a player.
     * @param user The user who's player to get.
     * @return The player or null if no player is found.
     */
    operator fun get(user: User) : P? = this.players[user.idLong]

    /** @return true if the game has a dev as a player */
    fun hasDev() = players.count { DEV_IDS.contains(it.key) } != 0

}
